docs/README-node-utils-APIs-xr.md

```diff
@@ -0,0 +1,277 @@
+'The command 'show running bgp all' was rejected with error:
+
+>>>show-cmd: show running bgp all <<<
+show running bgp all
+              ^
+% Invalid input detected at '^' marker.'
+```
+
+To correct this problem, the following changes were made in bgp_neighbor.yaml:
+
+```
+  _template:
+-   get_command: "show running router bgp"
++   ios_xr:
++     get_command: "show running router bgp"
++   nexus:
++     get_command: "show running bgp all"
+    context:
+      - "router bgp <asnum>"
+      - "(?)vrf <vrf>"
+      - "(?)neighbor <nbr>"
+```
+
+Here, the "global" get_command parameter was converted into two platform-specific parameters, while the context remained the same.
+
+#### <a name="get_value">get\_value/set\_value differences</a>
+
+Similar to the get_command differences above, the feature might have an attribute that is functionally equivalent on Nexus and XR, but have small syntactic differences.  An example of this for bgp neighbor is the "connected-check" attribute which produced the following error on XR:
+
+```bash
+  1) Error:
+TestRouterBgpNeighbor#test_connected_check:
+Cisco::CliError: CliError: '(unknown, see error message)' rejected with message:
+'The command '(unknown, see error message)' was rejected with error:
+
+Unable to process cmd, ret-val: 3, cmd: no router bgp 55 neighbor 1.1.1.1 disable-connected-check'
+```
+
+And required the following change in bgp\_neighbor.yaml:
+
+```
+  connected_check:
+    auto_default: false
+    kind: boolean
+-   get_value: '/^disable-connected-check$/'
+-   set_value: '<state> disable-connected-check'
++   ios_xr:
++     get_value: '/^ignore-connected-check$/'
++     set_value: '<state> ignore-connected-check'
++   nexus:
++     get_value: '/^disable-connected-check$/'
++     set_value: '<state> disable-connected-check'
+    default_value: true
+```
+
+#### <a name="unsupported">Unsupported attributes</a>
+
+Sometimes a Nexus command does not have an equivalent command on XR, and so the attribute will be considered unsupported on IOS XR.  An example of this is the bgp neighbor "low-memory-exempt" attribute which produced the now-familiar "unable to process cmd" error on XR.  The changes needed to mark this attribute as unsupported on XR are as follows:
+
+First, we exclude the attribute in bgp_neighbor.yaml:
+
+```
+  low_memory_exempt:
++   _exclude: [ios_xr]
+    kind: boolean
+    get_value: '/^low-memory exempt$/'
+    set_value: '<state> low-memory exempt'
+    default_value: false
+```
+
+This will cause any calls to config_set for this attribute to raise a Cisco::UnsupportedError, and calls to config_get and config_get_default will return nil.  Because of this, we must update the test_bgp_neighbor.rb minitest to expect these conditions on XR (see [Development Best Practices: MT6](./README-develop-best-practices.md#mt6)):
+
+```ruby
+  def test_low_memory_exempt
+    %w(default test_vrf).each do |vrf|
+      neighbor = create_neighbor(vrf)
++     if platform == :ios_xr
```

Comment:
Out of scope for this PR perhaps - but it should be really trivial for us to create a helper function in `ciscotest.rb` for this test - something like:

```ruby
assert_unsupported(object, property_name)
  assert_nil(object.send(property_name))
  assert_nil(object.send("default_#{property_name}"))
  assert_raises(Cisco::UnsupportedError) { object.send("#{property_name}=", true) }
end
```